binders/kafka-binder/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -502,6 +503,14 @@ public void fromHeaders(MessageHeaders headers, Headers target) {
 
 		}
 		handler.setHeaderMapper(mapper);
+
+		if (this.configurationProperties.isEnableObservation() || producerProperties.getExtension().isEnableObservation()) {
+			kafkaTemplate.setObservationEnabled(true);
+		}
+		kafkaTemplate.setApplicationContext(getApplicationContext());
+		customizeProducerMessageHandler(handler, destination.getName());
```

Comment:
This will cause it to be customized twice, right?

Reply:
Perhaps override it, call super, then aSI().

Reply:
Ya, from what I could see, it is an idempotent invocation. Do you see any issues with calling it twice? This way, we can call the `afterSingletonsInstantiated` from `KMCB` properly. If the user decides to enable observation using the customizer. , we will miss the `enableObservation` check in `KafkaTempate` `aSI` method without calling the customizer here. We can't call `aSI` on `KafkaTemplate` in `AMCB`.

Reply:
> Perhaps override it, call super, then `handler.getKT().aSI()`.

Done.

Reply:
Just to close the loop. Yes, it is likely idempotent, but we have no idea what user code might do. It might mutate objects outside of the template in a non-idempotent manner.